### PR TITLE
more explanation in complexity section

### PR DESCRIPTION
**Pull request by petrelharp**
This is following on from the discussion in #70.  The complexity section may be a bit wordy now, but should be closer to understandable.  I'm citing Wakeley, even though it's out of print, because it's the best explanation, I think.

### CODE REVIEW

**Comment by petrelharp** [comment]
oh, dammit, I haven't removed the `2`s.  Doing this.

File: forwards_paper.tex [hunk 1]
```diff
@@ -854,35 +871,45 @@ \subsubsection*{Computational complexity}
...
-have around $n(T)$ roots, where $n(T) \approx 1 + 2N/(T+2)$.
+How many of the resulting $2NT$ edges are required after simplification?
+As above, the expected number of edges is bounded by $N-2$
+plus four times the sum of the lengths of all marginal trees,
+because new edges appear with most recombinations that occur on any marginal tree.
+Therefore, we need to compute the mean total length of all branches in a coalescent tree
+up until time $T$ (or the common ancestor, whichever comes first).
+Since $k$ lineages coalesce at rate $\binom{k}{2}/N$,
+the expected length of time for which a marginal tree has $k$ branches is
+$N/(k(k-1)) = N(1/(k-1) - 1/k)$ generations.
```
Comment (jeromekelleher) [inline forwards_paper.tex]: Shouldn't this be 2N / (k(k - 1)), since \binom{k}{2} = k (k - 1) / 2?

File: forwards_paper.tex [hunk 2]
```diff
@@ -854,35 +871,45 @@ \subsubsection*{Computational complexity}
...
+Since there are $k$ branches for this time,
+this period is expected to contribute $N/(k-1)$ generations to the total tree length.
```
Comment (jeromekelleher) [inline forwards_paper.tex]: I don't follow the argument here --- what does total tree length have to do with it? I thought we were talking about time rather than total tree length?

File: forwards_paper.tex [hunk 3]
```diff
@@ -854,35 +871,45 @@ \subsubsection*{Computational complexity}
...
+By summing over $n < k \le N$,
```
Comment (jeromekelleher) [inline forwards_paper.tex]: I get: \sum_{k = n +1}^{N} 1 / (k - 1) - 1 / k = 1 / n - 1 / N; shouldn't this then be N / n - 1 generations ago?

File: forwards_paper.tex [hunk 4]
```diff
@@ -854,35 +871,45 @@ \subsubsection*{Computational complexity}
...
-This implies that after running for time $T$ we expect to
-have around $n(T)$ roots, where $n(T) \approx 1 + 2N/(T+2)$.
+How many of the resulting $2NT$ edges are required after simplification?
+As above, the expected number of edges is bounded by $N-2$
+plus four times the sum of the lengths of all marginal trees,
+because new edges appear with most recombinations that occur on any marginal tree.
+Therefore, we need to compute the mean total length of all branches in a coalescent tree
+up until time $T$ (or the common ancestor, whichever comes first).
+Since $k$ lineages coalesce at rate $\binom{k}{2}/N$,
+the expected length of time for which a marginal tree has $k$ branches is
```
Comment (jeromekelleher) [inline forwards_paper.tex]: Shouldn't this be k roots rather than branches? We have k extant lineages/roots in the tree and they coalesce at rate k choose 2, but the number of branches in the marginal tree is more complicated, isn't it?

**Review by jeromekelleher — COMMENTED** [review]
This is great, thanks for taking the time to do this @petrelharp. I understand the basic argument now, but I'm  still hazy on a few of the details. I've asked a bunch of questions that show I don't really know that much about coalescent theory!

Apologies for being annoying here and thanks for your patience...

**Comment by petrelharp** [comment]
Thanks for forcing me to actually explain this bit.  I was being lazy.

File: forwards_paper.tex (hunk 4 above)
Comment (petrelharp) [reply forwards_paper.tex]: I mean that if you are following the coalescent process up, then $k$ lineages coalesce at rate $k$ choose 2.  But then I waffled on whether to call these things "lineages" or "roots" or "branches".  I think that "lineages" is best, perhaps.  But then, we haven't explained what a lineage is, or indeed what "coalesce at rate" means.  I guess my goal here was to make this understandable to someone with a basic understanding of coalescent theory, in which case perhaps we don't need to define lineage.  What do you think about that strategy?  Options:

1. assume basic understanding of coal theory, along current lines
2. not assume this, and insert more explanation
3. not assume this and remove most of the arguments, just referring to "coalescent theory calculations tells us that X"

File: forwards_paper.tex (hunk 1 above)
Comment (petrelharp) [reply forwards_paper.tex]: oh.  yes.

File: forwards_paper.tex (hunk 2 above)
Comment (petrelharp) [reply forwards_paper.tex]: It is total tree length that tells us about the number of edges, because edges ~ recombs ~ total tree length. I tried to lay this out in the first paragraph; I should refer back to it better.  First we had to work out how (number of roots) relates to time because it's esaier to work out total tree length up until a certain number of roots than to work out total tree length up to a certain time.

File: forwards_paper.tex (hunk 3 above)
Comment (petrelharp) [reply forwards_paper.tex]: Er, let's see: the sum telescopes, so
```
 \sum_{k = n +1}^{N} (1 / (k - 1) - 1 / k) = (1/n - 1/(n-1)) + (1/(n-1) - 1/(n-2)) + ... + (1/(N-2) - 1/(N-1)) + (1/(N-1) - 1/N)
  = 1/n - 1/N
```
as you have.

So, yes, I agree. =)

File: forwards_paper.tex (hunk 4 above)
Comment (petrelharp) [reply forwards_paper.tex]: Actually, reading this again, I think it's right as stated: we're talking about the period of time over which there are `k` branches in the tree (thinking about the tree, not the coalescent process).  To figure this out, we ask about the time that the coalescent process has `k` lineages.  The questions there stand, though.

Comment (jeromekelleher) [reply forwards_paper.tex]: OK cool. I'd vote for option (1)

File: forwards_paper.tex (hunk 2 above)
Comment (jeromekelleher) [reply forwards_paper.tex]: I followed the bit about counting roots and computing the total tree length, but what threw me was the switching back and forth between the two. I'd be easier if we first computed the number of roots, and then stated the total tree length I think.

File: forwards_paper.tex [hunk 5]
```diff
@@ -854,35 +871,45 @@ \subsubsection*{Computational complexity}
...
+$N/(k(k-1)) = N(1/(k-1) - 1/k)$ generations.
+Since there are $k$ branches for this time,
+this period is expected to contribute $N/(k-1)$ generations to the total tree length.
+By summing over $n < k \le N$,
+the $N$ tips of a tree are expected to descend from only $n$ lineages
+around $N(1/(n-1) - 1/(N-1))$ generations ago.
+Inverting this relationship between time and number of roots
+implies that a marginal tree cut $T$ units of time ago
+is expected to have around $n(T)$ roots, where $n(T) \approx 1 + N/(T+2)$.
```
Comment (jeromekelleher) [inline forwards_paper.tex]: I just remembered this: n(T) is a bit confusing, as we're using n as something else just above. How about r(T) or something for the number of roots?

File: forwards_paper.tex (hunk 2 above)
Comment (petrelharp) [reply forwards_paper.tex]: gotcha; I can rearrange.

**Comment by petrelharp** [comment]
I've rewritten this bit substantially, to flow better and include more explanation.  In doing this, I could not reconstruct how I got to the previous bound: too bad, because it was a bit closer to the simulations, but I'm forced to conclude that it was in error.  We should expect substantial discrepancy between coalescent calculations and the whole-population genealogy, so I'm not worried about the current gap.  The form of the dependency (`T log N`) is right.

**Comment by petrelharp** [comment]
Possibly in bad form, I've merged to master here, but am leaving this open for discussion.

File: forwards_paper.tex [hunk 6]
```diff
@@ -816,80 +816,134 @@ \subsubsection*{Sequential simplification and prior history}
...
-($T = 20N$, say, giving a total space requirement of $60N^2$).
+$NT$ nodes and $2NT$ edges.
 After simplification, we are left with the tree sequence describing the history
-of only the current generation of $N$ individuals.
-If no coalescent events involve more than two lineages, this tree sequence has $2N-2$ edges to describe the leftmost tree;
-and each time the marginal tree changes along the sequence,
+of only the current generation of $N$ individuals, back to either $T$ generations ago,
+or their common ancestor, whichever comes first.
+The tree sequence must store edges describing the leftmost marginal tree,
+which requires at most $N-2$ edges.
```
Comment (jeromekelleher) [inline forwards_paper.tex]: 2N - 2, right?

File: forwards_paper.tex [hunk 7]
```diff
@@ -816,80 +816,134 @@ \subsubsection*{Sequential simplification and prior history}
...
+around $2(N/n - 1)$ generations ago.
+Inverting this relationship between time and number of roots
+implies that a marginal tree cut $T$ units of time ago
+% since 2N(1/n - 1/N) = 2(N/n-1) = T implies
+%    N/n = (T+2)/2 and so n = 2N/(T+2)
+is expected to have around $r(T)$ roots, where $r(T) \approx 2N/(T+2)$.
 The total tree length over this time is
-$2 N \sum_{k=n(T)+1}^{N-1} 1/k$, which
-% is approximately
+$2N \sum_{k=N}^{r(T)+1} 1/(k-1) \approx 2 N \log(N/r(T))$.
```
Comment (jeromekelleher) [inline forwards_paper.tex]: Two issues here:
1. Are the upper and lower bounds mixed up? r(T) can't be > N.
2. More pedantically, r(T) isn't an integer, so is this a floor or a ceiling?

File: forwards_paper.tex [hunk 8]
```diff
@@ -816,80 +816,134 @@ \subsubsection*{Sequential simplification and prior history}
...
+%     2N - 2 + 8 N \left\{ \log(N-1) - \log(r(T)) \}
+%        = 2N - 2 + 8 N \log( (T+2)(N-1)/2N )
+%        \approx 2 N (1 + 4 \log( (T+2)/2 ))
 % \end{align*}
-leads to an upper bound on the number of edges of
+Nonetheless, this leads as above to an upper bound on the number of edges of
 \begin{align}
     \label{eqn:edge_bound}
-    2 N \left( 1 + 4 \log\left( \frac{NT}{T + 2 N} \right)\right) .
+    2 N \left( 1 + 4 \log\left( \min\left(N, \frac{T+2}{2}\right) \right)\right) .
```
Comment (jeromekelleher) [inline forwards_paper.tex:940]: This makes sense based on the above, but isn't what we're plotting now.

File: forwards_paper.tex [hunk 9]
```diff
@@ -816,80 +816,134 @@ \subsubsection*{Sequential simplification and prior history}
...
+Since the tree has $k$ branches over this period,
+it is expected to contribute $2N/(k-1)$ to the total tree length.
+By summing over $n < k \le N$,
+the $N$ tips of a tree are expected to descend from only $n$ lineages
+around $2(N/n - 1)$ generations ago.
+Inverting this relationship between time and number of roots
+implies that a marginal tree cut $T$ units of time ago
+% since 2N(1/n - 1/N) = 2(N/n-1) = T implies
+%    N/n = (T+2)/2 and so n = 2N/(T+2)
+is expected to have around $r(T)$ roots, where $r(T) \approx 2N/(T+2)$.
```
Comment (jeromekelleher) [inline forwards_paper.tex]: Why the \approx here? We've solved exactly so far, haven't we? (Sure the whole calculation is quite approximate, but this is captured by the 'around' earlier)

**Review by jeromekelleher — COMMENTED** [review]
I really like the new presentation, it's great. I've pointed out a couple of niggling issues with the bound.

File: forwards_paper.tex (hunk 6 above)
Comment (petrelharp) [reply forwards_paper.tex]: whoops, yes. =)

File: forwards_paper.tex (hunk 9 above)
Comment (petrelharp) [reply forwards_paper.tex]: ah sure, good point.

File: forwards_paper.tex (hunk 7 above)
Comment (petrelharp) [reply forwards_paper.tex]: 1. Well,  I wrote it this way because it seems more natural to sum from now back until time T-ago; but good point, this might be confusing.
2. Yes, but let's sweep this under the rug, since then we change `\sum_{k=1}^x 1/k` to `log(x)` and it'd be nice to not propagate the floor.

File: forwards_paper.tex (hunk 8 above)
Comment (petrelharp) [reply forwards_paper.tex:940]: It's not?  Oh, right I [forgot to take out that `- 1/N`](https://github.com/petrelharp/ftprime_ms/blob/dc57882a131e038498ff360b4ef6b356e202db9d/sims/space-complexity.py#L68). Besides that, it agrees, no?

**Comment by petrelharp** [comment]
Thanks for the comments.  I've made changes for all of them.  Except the one about `floor`.

File: forwards_paper.tex (hunk 7 above)
Comment (jeromekelleher) [reply forwards_paper.tex]: WRT to (2), basically I was trying to reproduce your log approximation here with wolfram alpha and not getting anywhere. Looking at it again though, I'm assuming your using some well-know result about \sum_x^y 1/(k - 1)  \approx log(x / (y - 1)}?

If so, great, I'm happy not to think about it!

File: forwards_paper.tex (hunk 8 above)
Comment (jeromekelleher) [reply forwards_paper.tex:940]: Matches now, ta!

**Comment by jeromekelleher** [comment]
> Thanks for the comments. I've made changes for all of them. Except the one about floor.

LGTM. I'm sure the one about the floor is irrelevant. Merge away.

File: forwards_paper.tex (hunk 7 above)
Comment (petrelharp) [reply forwards_paper.tex]: Since
```
\sum_{k=1}^n 1/k \approx \log(n) + \gamma
```
then
```
\sum_{k=x}^y 1/k = \sum_{k=1}^x 1/k - \sum_{k=1}^{y-1} 1/k
  \approx \log(x) + \gamma - \log(y) - \gamma
  = \log(x/y)
```
I'll add a line of explanation (without the `\gamma`).

Comment (petrelharp) [reply forwards_paper.tex]: added small note here

File: forwards_paper.tex [hunk 10]
```diff
@@ -816,80 +816,135 @@ \subsubsection*{Sequential simplification and prior history}
...
+Inverting this relationship between time and number of roots
+implies that a marginal tree cut $T$ units of time ago
+% since 2N(1/n - 1/N) = 2(N/n-1) = T implies
+%    N/n = (T+2)/2 and so n = 2N/(T+2)
+is expected to have around $r(T)$ roots, where $r(T) = 2N/(T+2)$.
 The total tree length over this time is
-$2 N \sum_{k=n(T)+1}^{N-1} 1/k$, which
-% is approximately
+$2N \sum_{k=r(T)+1}^N 1/(k-1) \approx 2 N \log(N/r(T))$,
+since $\sum_{k=1}^x \approx \log(x) + \gamma$.
```
Comment (jeromekelleher) [inline forwards_paper.tex]: Missing the 1 / k argument to the sum here I think?

File: forwards_paper.tex (hunk 7 above)
Comment (jeromekelleher) [reply forwards_paper.tex]: Ah, that makes sense, very nice.

File: forwards_paper.tex (hunk 10 above)
Comment (petrelharp) [reply forwards_paper.tex]: er, yes, heh. =)

**Comment by petrelharp** [comment]
I've merged this manually.
